gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,271 @@
+
+    /* Extract which splits are not cleared and compute the amount we have to clear */
+    for (GList *node = xaccAccountGetSplitList (account); node; node = node->next)
+    {
+        auto split = static_cast<Split*> (node->data);
+
+        if (xaccSplitGetReconcile (split) != NREC)
+            toclear_value = gnc_numeric_sub_fixed
+                (toclear_value, xaccSplitGetAmount (split));
+        else if (gnc_numeric_zero_p (xaccSplitGetAmount (split)))
+            DEBUG ("skipping zero-amount split %p", split);
+        else if (gnc_numeric_check (xaccSplitGetAmount (split)))
+            DEBUG ("skipping invalid-amount split %p", split);
+        else if (end_date != INT64_MAX &&
+                 xaccTransGetDate (xaccSplitGetParent (split)) > end_date)
+            DEBUG ("skipping split after statement_date %p", split);
+        else
+            nc_vector.push_back (split);
+    }
+
+    if (gnc_numeric_zero_p (toclear_value))
+    {
+        g_set_error (error, autoclear_quark, AUTOCLEAR_NOP, "%s",
+                     _("Account is already at Auto-Clear Balance."));
+        goto skip_knapsack;
+    }
+    else if (nc_vector.empty ())
+    {
+        g_set_error (error, autoclear_quark, AUTOCLEAR_NOP, "%s",
+                     _("No uncleared splits found."));
+        goto skip_knapsack;
+    }
+
+    start_ticks = clock ();
+    sack.reserve(nc_vector.size() * 4);
+
+    for (auto& split : nc_vector)
+    {
+        auto amount = xaccSplitGetAmount (split);
+        std::vector<WorkItem> workvector {};
+
+        workvector.reserve (sack.size () + 1);
+        if (sack.find (amount) != sack.end ())
+            workvector.emplace_back (amount, SplitVec{});
+        else
+        {
+            SplitVec new_splits = { split };
+            workvector.emplace_back (amount, std::move(new_splits));
+        }
+
+        // printf ("new split. sack size = %ld\n", sack.size());
+
+        for (auto& [map_value, map_splits] : sack)
+        {
+            auto new_value = gnc_numeric_add_fixed (map_value, amount);
+            if (gnc_numeric_check (new_value))
+            {
+                g_set_error (error, autoclear_quark, AUTOCLEAR_OVERLOAD, "%s",
+                             "Overflow error: Amount numbers are too large!");
+                goto skip_knapsack;
+            }
+            if (map_splits.empty() ||
+                sack.find (new_value) != sack.end ())
+                workvector.emplace_back (new_value, SplitVec{});
+            else
+            {
+                SplitVec new_splits{};
+                new_splits.reserve(map_splits.size() + 1);
+                new_splits = map_splits;
+                new_splits.push_back (split);
+                workvector.emplace_back (new_value, std::move(new_splits));
+            }
+        }
+
+        for (auto& item : workvector)
+            sack.insert_or_assign(item.reachable_amount, std::move(item.splits_vector));
```

Comment:
this part *may* be `std::thread`able

Reply:
fwiw this part is technically parallelisable because all workitems (except possibly one -- the new split amount) have different reachable amounts, but all parallel workers would try to modify the same unordered_map, thus defeating the multithreading.